### ISSUE TITLE
query() - ignore nil/null/undefined

### ISSUE THREAD

**Issue by fabien** [dwstech/muster#31]
**Is your feature request related to a problem? Please describe.**

Currently, when a dynamic key/path returns `nil()`, the query will still try to extract the keys given:

```javascript
app.resolve(query(ref('users', 'unknown'), { id: true, firstName: true })).then(function(user) {
    console.log(user); // { id: undefined, firstName: undefined }
});
```
Is there a straightforward way to have it simply return the nil/null/undefined value itself in this case?

**Comment by dws-dave** [comment]
Hi @fabien, apologies for the lack of responses to your (excellent) questions at the moment - we're down a few staff for the Christmas break so aren't as responsive as we'd hope to be normally. We have just managed to get some [example projects](https://github.com/dwstech/muster-examples) uploaded though, which might provide you some usage examples while you wait for a proper answer. The RealWorld app also has a [live demo](https://infallible-chandrasekhar-152ee0.netlify.com/#/).

**Comment by fabien** [comment]
@dws-dave thanks for the update!

I discovered the example projects earlier today, quite helpful indeed!

Since I don't use React, I was wondering how the local container graph is actually implemented. I presume it is created and torn down at runtime, during the component life-cycle - is it a `scope` within the global graph? Or an entirely independent `muster` instance? I didn't figure out how to tear down a scope/muster instance (to free up resources) yet, so I couldn't really tell from looking at the source.

**Comment by dws-dave** [comment]
Hi @fabien. The `Provider` connects a top-level `muster` instance to a React app, and all the local container graphs are dynamically added and removed to it as they're instantiated or disposed. (This means you can't use local graphs without a top level `muster` instance in your `Provider` - although it can be completely empty.) Take a look at the [createContainer code](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts) for some pointers (`componentDidMount` and `componentWillUnmount` cover its lifecycle), I'll see if I can dig out some more meaningful documentation.

Are you considering any other view frameworks? We're hoping to get a POC implementation for Vue and Angular up and running over the next few months, but external contributions would be very welcome!

**Comment by fabien** [comment]
@dws-dave thanks! Could you tell me more on how to dynamically add/remove graphs to a top-level `muster`? Is it even possible to add/remove nodes at runtime, apart from having pre-defined slots as `variable` or `arrayList`?

I'm using the less main-stream https://riot.js.org/ as my view framework. I haven't started integrating `muster` with it, but I reckon it will be similar to how I work with Baobab.

**Comment by dws-dave** [comment]
I suppose the local graphs aren't actually inserted into the main graph - they're created as a [scope](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L88) with a reference to the main graph when the component is constructed (and disposed when the component is torn down), and only create their subscriptions when the respective components are mounted.

You can dynamically change the shape of a graph at runtime, although you'd need to use conditionals to do it - check out [this chunk](https://github.com/dwstech/muster-examples/blob/master/realworld-app/src/muster/api.ts#L325) in the RealWorld app which uses an `ifElse`. You can supply a radically different graph structure beneath each branch of the `ifElse` - Muster will cope perfectly happily - but you need to make sure the rest of you app will...

I've heard of Riot but never used it - we'll be very interested to hear how you get on!

(disclaimer: I may be wrong on some of the muster-react internal stuff here but we're never far away from someone who knows better)

**Comment by dws-dave** [comment]
@fabien That's an excellent question, and I'm afraid I can't give you a decent answer at this point 😓

The rest of the team will be back in the new year - we'll get you some proper answers then!

**Comment by dws-tomasz** [comment]
Hi @fabien. Going back to your original question, in the current version of Muster it is not possible to simply return null/undefined when a query resolves to nil(). At the moment the `query()` node preserves the shape of the query, and returns `undefined` for leaves. I'm sure it's possible to make this behavior configurable, for example:
```javascript
query(
  ref('users', 'unknown'),
  { id: true, firstName: true },
  { omitNils: true }
);
```
In your case the following query would resolve to `undefined`.

**Comment by dws-tomasz** [comment]
With regards to the question about how to dynamically add/remove graphs to Muster it's quite advanced topic, but I will try to give you as good explanation of this as I can ;)

Before we dig into the explanation I'd like to add a refresher of some core concepts.

Firstly, I'd like to mention is that there's no restriction on the node used as a target of a `query()` node. In most of Muster examples we've been using either `root()` or `ref()` nodes, but you can also resolve a following query:

```javascript
query(
  tree({
    firstName: 'Bob',
    lastName: 'Smith',
  }),
  {
    firstName: true,
    lastName: true,
  },
);
```

The target of the `query()` node could also be defined as a `scope()` node:

```javascript
query(
  scope({
    firstName: 'Bob',
    lastName: 'Smith',
  }),
  {
    firstName: true,
    lastName: true,
  },
);
```

Second thing I'd like to mention is that nodes located in the scope do not have direct access outside that scope. The only way the scoped nodes can access wider graph is by injecting nodes through context. For example, given an application:

```javascript
const app = muster({
  name: 'Name from root of the graph',
  scope: scope({
    name: 'Name from scope',
    getName: get(root(), 'name'), // ref('name')
  }),
});
```

Resolving `await app.resolve(ref('scope', 'getName'))` returns 'Name from scope'. This happens because the scope node overrides the value of a `root()` node, which essentially prevents the nodes from being able to access anything outside that scope.

The `scope()` node optionally takes context values, which can be injected into that scope. This allow creating links between scope and its parent scope. There's no restrictions on what that link points to, as it can even point to the original root of the parent scope:

```javascript
const app = muster({
  name: 'Name from root of the graph',
  scope: scope({
    name: 'Name from scope',
    getName: get(context('originalRoot'), 'name'),
  }, {
    originalRoot: root(),
  }),
});
```

In this case resolving `await app.resolve(ref('scope', 'getName'))` returns 'Name from root of the graph'.

One thing to remember is that in the prior example I used string keys for context values, but they can also be defined as symbols (something we use to our advantage in the `createContainerFactory()` function).

The final thing is that the `scope()` node definition exposes a `dispose()` function (nice work for spotting it!). As you might remember from Muster tutorials, certain Muster nodes can have their own state, and their state (usually) lives for as long as the node has open subscriptions. The reason why I wrote "usually" is because there's a cheeky `variable()` node, which actually ignores the number of open subscriptions, and stores the value for as long as the scope it lives in exist, or until the variable() is reset (using `reset()` node, or a `resetVariablesInScope()` event).

Imagine a scenario where a scope with a bunch of variables is created by the component, the variables are set to some value, then the component gets destroyed but the state of these variables remain. For this reason we've added a `dispose()` function to the `scope()` node definition. This function is used to force removal of the state of every child node and scope of that scope() node definition.

Right, now that we've got "basics" out of the way let's get into the implementation of the dynamic adding/removing graphs 😉

The simplified workflow looks like this:
1. Based on the `graph` property from an object passed into the `container()` function create a `scope()` node, while retaining a link to the root of the Muster graph through some context value [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L83)
2. Remember that `scope()` node definition, as it will be used to build both the `query()` node, and when component gets unmounted to dispose the retained nodes. [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L150)
3. Build a `query()` that uses a `scope()` node definition from the step 2., and with the fields coming from transformed `props` property (again, from an object passed into the `container()` function). [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L151)
4. Subscribe to that `query()` node, and render the component with the results from Muster. [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L217) and [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L244)
5. (when unmounting) Close `query()` subscription, and then dispose the `scope()` node definition using the `.dispose()` function. [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L373) and [link](https://github.com/dwstech/muster/blob/develop/packages/muster-react/src/utils/create-container-factory.ts#L357)

Let me know if you have something isn't clear 😉

**Comment by dws-tomasz** [comment]
FYI, that's why we've got the `global()` node in Muster-React. The implementation of that node boils down to accessing global root() node from the context. To illustrate this take a look at this example:
```javascript
// Create Muster app
const app = muster({
  user: {
    name: 'Bob',
  },
});

// Simple implementation of the global()
function global(...path) {
  return { root: context('myRoot'), path };
}

// Create component scope
const myScope = scope(
  toNode({
    name: ref(global('user', 'name')),
  },
  { myRoot: root() },
);

// Create component query that targets the component scope
const nameQuery = query(myScope, { name: true });

// Resolve the query
const subscription = app.resolve(nameQuery).subscribe((values) => { /* ... */ });

// And then component is disposed, time to do a cleanup
subscription.unsubscribe();
myScope.dispose();
```

**Comment by fabien** [comment]
> Hi @fabien. Going back to your original question, in the current version of Muster it is not possible to simply return null/undefined when a query resolves to nil(). At the moment the `query()` node preserves the shape of the query, and returns `undefined` for leaves. I'm sure it's possible to make this behavior configurable, for example:
> 
> ```js
> query(
>   ref('users', 'unknown'),
>   { id: true, firstName: true },
>   { omitNils: true }
> );
> ```
> 
> In your case the following query would resolve to `undefined`.

This option sounds really useful! Will you introduce it?

**Comment by fabien** [comment]
> With regards to the question about how to dynamically add/remove graphs to Muster it's quite advanced topic, but I will try to give you as good explanation of this as I can ;)

Thanks for the elaborate explanation. I think I understand it now! So you're not actually attaching the scope to the main tree, but simply query a scope instance. Which then brings in the main context as specified on scope creation. Interesting!

**Comment by dws-tomasz** [comment]
> This option sounds really useful! Will you introduce it?

Yup. You can expect this feature in the next few days 😉

**Comment by dws-dave** [comment]
Closed via #51
